File: shared/types/staff.ts
// 운영진 소개 API 계약 — FE↔BE 합의 파일
// 변경 시 양 팀 합의 필요 (shared/ 규칙)
//
// 이름 확정: Staff. 권한 레벨 admin(예: /api/admin/...)과 리소스명이 겹치는 걸 피하려고
// "운영진" 리소스는 Admin이 아니라 Staff로 부른다 (이슈 #87).

/** GET /api/staff — 공개 전체 목록, sortOrder 오름차순(동률 시 id 오름차순) */
export interface Staff {
  id: number;
  name: string;
  position: string;
  department: string;
  admissionYear: number;
  photoUrl: string;
  introduction: string | null;
  /** 운영진 활동 이력. 없으면 빈 배열. 랜딩 카드에 순서대로 노출된다. */
  activities: string[];
  sortOrder: number;
}

/**
 * POST /api/admin/staff — 관리자 전용.
 * phone은 저장만 하고 응답하지 않으며, publicationConsent=true이면 publicationConsentedAt도 함께 보낸다.
 */
export interface StaffCreateRequest {
  name: string;
  position: string;
  department: string;
  admissionYear: number;
  photoUrl: string;
  introduction?: string;
  activities?: string[];
  sortOrder: number;
  studentId?: string;
  phone?: string;
  publicationConsent?: boolean;
  publicationConsentedAt?: string;
}

/** PATCH /api/admin/staff/{id} — 관리자 전용. name·department·admissionYear는 등록 후 불변 */
export interface StaffUpdateRequest {
  position?: string;
  photoUrl?: string;
  introduction?: string;
  /** null이 아니라 값이 오면 활동을 통째로 교체한다. */
  activities?: string[];
  sortOrder?: number;
  studentId?: string;
  phone?: string;
  publicationConsent?: boolean;
  publicationConsentedAt?: string;
}

/** GET /api/admin/staff와 관리자 생성·수정 응답. phone은 저장만 하고 응답하지 않는다. */
export interface StaffAdminSummary extends Staff {
  studentId: string | null;
  publicationConsent: boolean;
  publicationConsentedAt: string | null;
}

File: backend/src/main/java/likelion/khu/website/staff/Staff.java
package likelion.khu.website.staff;

import jakarta.persistence.*;
import lombok.Getter;
import lombok.NoArgsConstructor;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

@Entity
@Table(name = "staff")
@Getter
@NoArgsConstructor
public class Staff {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(columnDefinition = "integer")
    private Long id;

    @Column(nullable = false)
    private String name;

    @Column(nullable = false)
    private String position;

    @Column(nullable = false)
    private String department;

    @Column(nullable = false)
    private Integer admissionYear;

    @Column(nullable = false)
    private String photoUrl;

    @Column(columnDefinition = "TEXT")
    private String introduction;

    // 운영진 소개 카드에 보이는 "활동 이력". member_roles와 같은 컬렉션 테이블 패턴.
    // @OrderColumn(sort_order)로 화면에 보이는 순서를 고정한다.
    @ElementCollection(fetch = FetchType.EAGER)
    @CollectionTable(name = "staff_activities", joinColumns = @JoinColumn(name = "staff_id", columnDefinition = "bigint"))
    @OrderColumn(name = "sort_order")
    @Column(name = "activity", nullable = false)
    private List<String> activities = new ArrayList<>();

    private String studentId;

    private String phone;

    @Column(nullable = false)
    private boolean publicationConsent;

    private LocalDateTime publicationConsentedAt;

    @Column(nullable = false)
    private Integer sortOrder;

    @Column(nullable = false)
    private String createdBy;

    @Column(nullable = false)
    private String updatedBy;

    @Column(nullable = false)
    private LocalDateTime createdAt;

    @Column(nullable = false)
    private LocalDateTime updatedAt;

    public static Staff create(String name, String position, String department, Integer admissionYear,
                               String photoUrl, String introduction, List<String> activities, Integer sortOrder,
                               String studentId, String phone, boolean publicationConsent,
                               LocalDateTime publicationConsentedAt, String createdBy) {
        Staff s = new Staff();
        s.name = name;
        s.position = position;
        s.department = department;
        s.admissionYear = admissionYear;
        s.photoUrl = photoUrl;
        s.introduction = introduction;
        s.activities = activities != null ? new ArrayList<>(activities) : new ArrayList<>();
        s.sortOrder = sortOrder;
        s.studentId = studentId;
        s.phone = phone;
        s.publicationConsent = publicationConsent;
        s.publicationConsentedAt = publicationConsentedAt;
        s.createdBy = createdBy;
        s.updatedBy = createdBy;
        LocalDateTime now = LocalDateTime.now();
        s.createdAt = now;
        s.updatedAt = now;
        return s;
    }

    public void update(String name, String position, String department, Integer admissionYear,
                       String photoUrl, String introduction, List<String> activities, Integer sortOrder,
                       String studentId, String phone, Boolean publicationConsent,
                       LocalDateTime publicationConsentedAt, String updatedBy) {
        if (name != null) this.name = name;
        if (position != null) this.position = position;
        if (department != null) this.department = department;
        if (admissionYear != null) this.admissionYear = admissionYear;
        if (photoUrl != null) this.photoUrl = photoUrl;
        if (introduction != null) this.introduction = introduction;
        if (activities != null) {
            this.activities.clear();
            this.activities.addAll(activities);
        }
        if (sortOrder != null) this.sortOrder = sortOrder;
        if (studentId != null) this.studentId = studentId;
        if (phone != null) this.phone = phone;
        if (publicationConsent != null) {
            this.publicationConsent = publicationConsent;
            this.publicationConsentedAt = publicationConsentedAt;
        }
        this.updatedBy = updatedBy;
        this.updatedAt = LocalDateTime.now();
    }
}
